Encapsulate token request exchange between client and issuer (#199)
tfpauly/privacy-proxy: draft-privacypass-rate-limit-tokens.md

1. Issuer Request URI: a token request URL for generating access tokens.
   For example, an Issuer URL might be https://issuer.example.net/token-request. This parameter
   uses resource media type "text/plain".
1. Issuer Encapsulation Key: a `EncapsulationKey` structure as defined below to use when encapsulating

>    information to the Issuer in issuance requests. This parameter uses resource media type

Reviewer: ```suggestion
   information, such as the origin name, to the Issuer in issuance requests. This parameter uses resource media type
```

    Sec-Token-Request-Blind = sf-binary
~~~

The "Sec-Token-Request-Key" is an Item Structured Header {{!RFC8941}}. Its
value MUST be a Byte Sequence. This header is sent on Client-to-Attester
requests ({{request-one}}), and contains a per-request public key.
Its ABNF is:

~~~

>     Sec-Token-Request-Key = sf-binary

Reviewer: I think we need to add this to IANA considerations?

Author: Yep! Forgot to do that.

The Client then uses Client Key to generate its one-time-use request public
key `request_key` and blind `request_blind` as described in {{client-anon-issuer-origin-id}}.

> The Client then constructs a OriginTokenRequest value, denoted `origin_token_request`,

Reviewer: ```suggestion
The Client then constructs a InnerTokenRequest value, denoted `inner_token_request`,
```

combining `blinded_msg`,  `request_key`, and the origin name as follows:

~~~
struct {
  uint8_t blinded_msg[Nk];
  uint8_t request_key[49];
  uint8_t padded_origin_name<0..2^16-1>;

> } OriginTokenRequest;

Reviewer: ```suggestion
} InnerTokenRequest;
```

generated as SHA256(public_key), where public_key is a DER-encoded SubjectPublicKeyInfo
object carrying Token Key.

- "issuer_encap_key_id" is a collision-resistant hash that identifies the Issuer
Encryption Key, generated as SHA256(EncapsulationKey).

> - "encrypted_token_request" is an encrypted structure that contains an OriginTokenRequest

Reviewer: ```suggestion
- "encrypted_token_request" is an encrypted structure that contains an InnerTokenRequest
```

Upon receipt of the forwarded request, the Issuer validates the following conditions:

- The TokenRequest contains a supported token_type
- The TokenRequest.token_key_id and TokenRequest.issuer_encap_key_id correspond to known
Token Keys and Issuer Encapsulation Keys held by the Issuer.
- The TokenRequest.encrypted_token_request can be decrypted using the
Issuer's private key (the private key associated with Issuer Encapsulation Key), and contains

> a valid OriginTokenRequest whose unpadded origin name matches an Origin Name that is served by
> the Issuer. The Origin name associated with the OriginTokenRequest value might be the empty string "",

Reviewer: ```suggestion
a valid InnerTokenRequest whose unpadded origin name matches an Origin Name that is served by
the Issuer. The Origin name associated with the InnerTokenRequest value might be the empty string "",
```

## Issuer-to-Attester Response {#response-one}

If the Issuer is willing to give a token to the Client, the Issuer decrypts

> TokenRequest.encrypted_token_request to discover a OriginTokenRequest value. If this fails,

Reviewer: ```suggestion
TokenRequest.encrypted_token_request to discover a InnerTokenRequest value. If this fails,
```

the Issuer rejects the request with a 400 error. Otherwise, the Issuer validates and
processes the token request with Issuer Origin Secret corresponding to the designated
Origin as described in {{issuer-anon-issuer-origin-id}}. If this fails, the Issuer
rejects the request with a 400 error. Otherwise, the output is
index_key.

The Issuer completes the issuance flow by computing a blinded response as follows:

~~~

> blind_sig = rsabssa_blind_sign(skP, OriginTokenRequest.blinded_msg)

Reviewer: ```suggestion
blind_sig = rsabssa_blind_sign(skP, InnerTokenRequest.blinded_msg)
```

Issuers decrypt the token request using their corresponding private key. This process yields
the decrypted token request as well as a shared encryption context between Client and Issuer.
Issuers encapsulate their token response to the Client using an ephemeral key derived from this
shared encryption context. This process ensures that the Attester learns neither the token
request or response information.

Client to Issuer encapsulation is described in {{encap-client-to-issuer}}, and Issuer to
Client encapsulation is described in {{encap-issuer-to-client}}.

## Client to Issuer Encapsulation {#encap-client-to-issuer}

Given a `EncapsulationKey` (Issuer Encapsulation Key), Clients produce encrypted_token_request
using the following values:

- the one octet key identifier from the Name Key, keyID, with the corresponding KEM identified by kemID,
the public key from the configuration, pkI, and;
- a selected combination of KDF, identified by kdfID, and AEAD, identified by aeadID.

Beyond the key configuration inputs, Clients also require the following inputs defined
in {{request-one}}: `token_key_id`, `blinded_msg`, `request_key`, `origin_name`, and
`issuer_encap_key_id`.

> Together, these are used to encapsulate an OriginTokenRequest and produce an encrypted token

Reviewer: ```suggestion
Together, these are used to encapsulate an InnerTokenRequest and produce an encrypted token
```

request (`encrypted_token_request`).

`origin_name` contains the name of the origin that initiated the challenge, as
taken from the TokenChallenge.origin_info field. If the TokenChallenge.origin_info field
is empty, `origin_name` is set to the empty string "".

The process for generating `encrypted_token_request` from `blinded_msg`, `request_key`, and
`origin_name` values is as follows:

1. Compute an {{HPKE}} context using pkI, yielding context and encapsulation key enc.
1. Construct associated data, aad, by concatenating the values of keyID, kemID, kdfID,
   aeadID, and all other values of the TokenRequest structure.
1. Pad origin_name with N zero bytes, where N = 31 - ((L - 1) % 32) and L is the length
   of origin_name. If L is 0, N = 32. Denote this padding process as the function `pad`.
1. Encrypt (seal) the padded origin_name with aad as associated data using context, yielding ciphertext ct.
1. Concatenate the values of aad, enc, and ct, yielding encrypted_token_request.

Note that enc is of fixed-length, so there is no ambiguity in parsing this structure.

In pseudocode, this procedure is as follows:

~~~

> enc, context = SetupBaseS(pkI, "OriginTokenRequest")

Reviewer: ```suggestion
enc, context = SetupBaseS(pkI, "TokenRequest")
```

aad = concat(encode(1, keyID),
             encode(2, kemID),
             encode(2, kdfID),
             encode(2, aeadID),
             encode(2, token_type),
             encode(1, token_key_id),
             encode(32, issuer_encap_key_id))
padded_origin_name = pad(origin_name)
input = concat(encode(Nk, blinded_msg),
               encode(49, request_key),
               encode(len(padded_origin_name), padded_origin_name))
ct = context.Seal(aad, input)
encrypted_token_request = concat(enc, ct)
~~~

> Issuers reverse this procedure to recover the OriginTokenRequest value by computing the AAD as
> described above and decrypting encrypted_token_request with their private key skI (the private
> key corresponding to pkI). The `origin_name` value is recovered from OriginTokenRequest.padded_origin_name
> by stripping off padding bytes. In pseudocode, this procedure is as follows:

Reviewer: ```suggestion
Issuers reverse this procedure to recover the InnerTokenRequest value by computing the AAD as
described above and decrypting encrypted_token_request with their private key skI (the private
key corresponding to pkI). The `origin_name` value is recovered from InnerTokenRequest.padded_origin_name
by stripping off padding bytes. In pseudocode, this procedure is as follows:
```

~~~
enc, ct = parse(encrypted_token_request)
aad = concat(encode(1, keyID),
             encode(2, kemID),
             encode(2, kdfID),
             encode(2, aeadID),
             encode(2, token_type),
             encode(1, token_key_id),
             encode(32, issuer_encap_key_id))
context = SetupBaseR(enc, skI, "TokenRequest")
origin_token_request, error = context.Open(aad, ct)
~~~

> The `OriginTokenRequest.blinded_msg` and `OriginTokenRequest.request_key` values, along

Reviewer: ```suggestion
The `InnerTokenRequest.blinded_msg` and `InnerTokenRequest.request_key` values, along
```

with the unpadded `origin_name` value, are used by the Issuer as described in {{request-two}}.

## Issuer to Client Encapsulation {#encap-issuer-to-client}

Given an HPKE context `context` computed in {{encap-client-to-issuer}}, Issuers encapsulate
their token response `blind_sig`, yielding an encrypted token response `encrypted_token_response`,
to the Client as follows:

> 1. Export a secret `secret` from `context`, using the string "OriginTokenResponse" as context.

Reviewer: ```suggestion
1. Export a secret `secret` from `context`, using the string "TokenResponse" as context.
```

   key and nonce associated with `context`.

2. Generate a random value of length `max(Nn, Nk)` bytes, called `response_nonce`.

3. Extract a pseudorandom key `prk` using the `Extract` function provided by
   the KDF algorithm associated with `context`. The `ikm` input to this
   function is `secret`; the `salt` input is the concatenation of `enc` (from
   `enc_request`) and `response_nonce`

4. Use the `Expand` function provided by the same KDF to extract an AEAD key
   `key`, of length `Nk` - the length of the keys used by the AEAD associated
   with `context`. Generating `key` uses a label of "key".

5. Use the same `Expand` function to extract a nonce `nonce` of length `Nn` -
   the length of the nonce used by the AEAD. Generating `nonce` uses a label of
   "nonce".

6. Encrypt `blind_sig`, passing the AEAD function Seal the values of `key`,
   `nonce`, empty `aad`, and a `pt` input of `request`, which yields `ct`.

7. Concatenate `response_nonce` and `ct`, yielding an Encapsulated Response
   `enc_response`. Note that `response_nonce` is of fixed-length, so there is no
   ambiguity in parsing either `response_nonce` or `ct`.

In pseudocode, this procedure is as follows:

~~~

> secret = context.Export("OriginTokenResponse", Nk)

Reviewer: ```suggestion
secret = context.Export("TokenResponse", Nk)
```

- origin_name: The Origin Name to encrypt, represented as a hexadecimal string.
- kem_id, kdf_id, aead_id: The HPKE algorithms comprising the ciphersuite DHKEM(X25519, HKDF-SHA256), HKDF-SHA256, AES-128-GCM.
- issuer_encap_key_seed: The seed used to derive the private key corresponding to
  Issuer Encapsulation Key via the DeriveKeyPair function as defined in {{Section 7.1.3. of HPKE}},
  represented as a hexadecimal string.
- issuer_encap_key: The public Issuer Encapsulation Key, represented as a hexadecimal string.
- token_type: The type of the protocol specified in this document, 0x0003.
- token_key_id: The ID of Token Key computed as in {{request-one}}, a single octet.
- blinded_msg: A random blinded_msg value, represented as a hexadecimal string.
- request_key: A random request_key value, represented as a hexadecimal string.
- issuer_encap_key_id: The Issuer Encapsulation Key ID computed as in {{request-one}}, represented as a hexadecimal string.

> - encrypted_token_request: The encrypted OriginTokenRequest, represented as a hexadecimal string.

Reviewer: ```suggestion
- encrypted_token_request: The encrypted InnerTokenRequest, represented as a hexadecimal string.
```

44c4ec6abb1c846a16a222116ad94
request_key: 8a9ffd91064e498f0b14b4b82d65255b2859da8d349e2bb362ca7c8211e
23fcd5cd01813ef6eb4b994d532c3520e1aa89e
issuer_encap_key_id:
590ef89ec14ebd7cd0598c7c1e045838e7c484713265c10d923d314fa1ef433f

> encrypted_token_request: b9d946ee16d9114b6d42524a1fe1e5fae9768c45d4b48ce13

Reviewer 2: This seems not to have been updated. Also we should delete the "blinded_msg" above, and add the issuer response?

Author: Yeah, I haven't updated any of the test vectors yet, but will do so before I mark this PR as ready. This particular change is the result of s/encrypted_origin_name/encrypted_token_request.
